BranchSDK/tools/build_wix.py

```diff
@@ -0,0 +1,304 @@
+#! /usr/bin/env python
+
+"""
+Script to build Product.wxs for Wix installer using the output
+of stage.py under build\*\stage.
+"""
+
+# TODO: Don't necessarily need ET here, since we're just generating
+# text, but this may be more flexible.
+
+import json, os, shutil
+from uuid import uuid4
+from xml.dom import minidom
+import xml.etree.ElementTree as ET
+from xml.etree.ElementTree import Comment, Element, SubElement
+
+# -----
+# ----- Utilities used below
+# -----
+
+# https://pymotw.com/2/xml/etree/ElementTree/create.html
+def prettify(elem):
+    """Return a pretty-printed XML string for the Element.
+    """
+    rough_string = ET.tostring(elem, 'utf-8')
+    reparsed = minidom.parseString(rough_string)
+    return reparsed.toprettyxml(indent="  ")
+
+# Method used to generate a unique identifier for a file or directory.
+def file_identifier(path):
+    build_root = os.path.abspath(os.path.dirname(__file__) + "../../build")
+    # Wix error message:
+    #  Identifiers may contain ASCII characters A-Z, a-z, digits,
+    #  underscores (_), or periods (.).  Every identifier must
+    #  begin with either a letter or an underscore.
+    return path.replace(build_root + os.sep, "").replace(os.sep, ".")
+
+# Find all subdirectories to all depths. Returns a flat list, for
+# a ComponentGroup.
+def all_subdirs(path):
+    if not os.path.isdir(path):
+        return []
+
+    # Now path must be a directory
+
+    # List of full paths to all immediate children of path
+    all_paths = [os.path.join(path, f) for f in os.listdir(path)]
+
+    # Filter out only directories.
+    # List of full paths to all immediate subdirectories of path
+    all_dirs = [f for f in all_paths if os.path.isdir(f)]
+
+    # Now all_nested is a list of lists. Flatten it out. There's no
+    # possibility of repetition of elements.
+    result = all_dirs
+    for d in all_dirs:
+        nested = all_subdirs(d)
+        result = result + nested
+    return result
+
+# -----
+# ----- Conveniences to make specific ET elements
+# -----
+
+def make_directory_elem(elem, identifier, name=None):
+    if name:
+        return SubElement(elem, "Directory", {"Id": identifier, "Name": name})
+    else:
+        return SubElement(elem, "Directory", {"Id": identifier})
+
+def make_component_elem(elem, identifier, directory):
+    # TODO: Generating a new UUID each time here. If that's an issue,
```

Comment:
This means the script is not idempotent, which is a shame. It's probably a good idea to do what's described in the comment.

Reply:
This was a bit of a problem. If you didn't actually uninstall the previous build before testing out a new one, it would show up more than once under Add and Remove Programs. This has been fixed. The long-term solution should be to use the `heat.exe` tool from Wix, separate the project into a number of separate .wxs files and get rid of this Python script, but for now it does the job.